### PR TITLE
Allow to create regular apikeys for data observatory datasets

### PR DESCRIPTION
**Pull request by dgaubert**

### CODE REVIEW

File: spec/models/carto/api_key_spec.rb [hunk 1]
```diff
@@ -812,6 +823,30 @@ def validate_view_api_key(view_name, create_query, drop_query)
...
+        grants = [apis_grant(['maps']), data_observatory_datasets_grant]
+        expected = ['carto-do.here.pointsofinterest_pointsofinterest_usa_latlon_v1_quarterly_v1']
+        api_key = @carto_user1.api_keys.create_regular_key!(name: 'data-observatory', grants: grants)
+
+        api_key.should be
+        api_key.data_observatory_datasets.should eq expected
+        data_observatory_datasets = $users_metadata.hget(api_key.send(:redis_key), :data_observatory_datasets)
+        expect(JSON.parse(data_observatory_datasets, symbolize_names: true)).to eql(expected)
+
+        api_key.destroy
```
Comment (amiedes) [inline spec/models/carto/api_key_spec.rb:846]: Ideally the tests should do the cleanup themselves, so we avoid having to write and maintain code explicitly to doing the teardown. For things like `api_key.destroy` is not that severe, but we have really messed up teardowns in our codebase related to users and visualizations removal.

Ruby has the [database cleaner](https://github.com/DatabaseCleaner/database_cleaner) library, but I think it's not configured correctly.

This is a bit opinionated, but I'd prefer to remove the `api_key.destroy` if you see the tests keep passing (but if you see removing it breaks something else you can leave it).

File: spec/requests/carto/api/api_keys_controller_spec.rb [hunk 2]
```diff
@@ -227,6 +229,37 @@ def user_req_params(user, token = nil)
...
+            ]
+          }
+        ]
+        name = 'foo'
+        payload = { name: name, grants: grants }
+        auth_user(@carto_user)
+        post_json api_keys_url, auth_params.merge(payload), auth_headers do |response|
+          response.status.should eq 201
+          api_key_response = response.body
+          api_key_response[:id].should_not be
```
Comment (amiedes) [inline spec/requests/carto/api/api_keys_controller_spec.rb:251]: Could you replace this shoulds with the expect alternative?

File: spec/requests/carto/api/api_keys_controller_spec.rb [hunk 3]
```diff
@@ -227,6 +229,37 @@ def user_req_params(user, token = nil)
...
+          api_key_response[:name].should eq name
+          api_key_response[:user][:username].should eq @carto_user.username
+          api_key_response[:type].should eq 'regular'
+          api_key_response[:token].should_not be_empty
+          expected_datasets_granted = ['carto-do.here.pointsofinterest_pointsofinterest_usa_latlon_v1_quarterly_v1']
+          request_datasets_granted = grants.find { |grant| grant[:type] == 'data-observatory' }[:datasets]
+          expect(request_datasets_granted).to eql(expected_datasets_granted)
+          Carto::ApiKey.where(name: api_key_response[:name]).each(&:destroy)
```
Comment (amiedes) [inline spec/requests/carto/api/api_keys_controller_spec.rb:259]: Same comment as in the other destroy

**Review by amiedes — CHANGES_REQUESTED** [review]
Nice work 💪. I've left some comments, but they should be straightforward to fix.

File: spec/models/carto/api_key_spec.rb [hunk 4]
```diff
@@ -1,3 +1,5 @@
+# rubocop:disable InstanceVariable, GlobalVars, DepartmentName
```
[the remark this answers is not in the document]
Comment (dgaubert) [reply spec/models/carto/api_key_spec.rb]: Done!

File: app/models/carto/api_key.rb [hunk 5]
```diff
@@ -455,6 +463,13 @@ def process_dataset_metadata_permissions
       dataset_metadata_grants.try(:[], :table_metadata)
     end
 
+    def process_data_observatory_datasets
+      data_observatory_grants = grants.find { |v| v[:type] == 'data-observatory' }
+      return nil if data_observatory_grants.blank?
```
[the remark this answers is not in the document]
Comment (dgaubert) [reply app/models/carto/api_key.rb]: Done!

File: spec/models/carto/api_key_spec.rb [hunk 6]
```diff
@@ -812,6 +823,30 @@ def validate_view_api_key(view_name, create_query, drop_query)
...
     end
 
+    describe 'data observatory datasets api key' do
+      before do
+        db_role = Carto::DB::Sanitize.sanitize_identifier("carto_role_#{SecureRandom.hex}")
+        described_class.any_instance.stubs(:db_role).returns(db_role)
+      end
+
+      after do
+        described_class.any_instance.unstub(:db_role)
```
[the remark this answers is not in the document]
Comment (dgaubert) [reply spec/models/carto/api_key_spec.rb]: Done!

File: spec/models/carto/api_key_spec.rb (hunk 1 above)
Comment (dgaubert) [reply spec/models/carto/api_key_spec.rb]: Done!

File: spec/requests/carto/api/api_keys_controller_spec.rb (hunk 3 above)
Comment (dgaubert) [reply spec/requests/carto/api/api_keys_controller_spec.rb]: Done!

File: spec/requests/carto/api/api_keys_controller_spec.rb (hunk 2 above)
Comment (dgaubert) [reply spec/requests/carto/api/api_keys_controller_spec.rb]: Done!

File: spec/models/carto/api_key_spec.rb [hunk 7]
```diff
@@ -812,6 +823,30 @@ def validate_view_api_key(view_name, create_query, drop_query)
...
+      end
+
+      it 'grants with data observatory datasets' do
+        grants = [apis_grant(['maps']), data_observatory_datasets_grant]
+        expected = ['carto-do.here.pointsofinterest_pointsofinterest_usa_latlon_v1_quarterly_v1']
+        api_key = @carto_user1.api_keys.create_regular_key!(name: 'data-observatory', grants: grants)
+
+        api_key.should be
+        api_key.data_observatory_datasets.should eq expected
```
[the remark this answers is not in the document]
Comment (dgaubert) [reply spec/models/carto/api_key_spec.rb]: Done!

**Review by amiedes — APPROVED** [review]
👍

**Comment by dgaubert** [comment]
Already tested in staging: 

Should work: https://maps-api-v2.carto-staging.com/user/juanra-do-org-dgaubert/tilejson/tileset/data-observatory/carto-do.experian.demographics_consumersegments_kor_grid_v1_yearly_2020?api_key=7d5KVB9CJuamjESX4vRW4Q

Should NOT work: https://maps-api-v2.carto-staging.com/user/juanra-do-org-dgaubert/tilejson/tileset/data-observatory/carto-do.experian.demographics_consumersegments_kor_grid_v1_yearly_2020?api_key=Dabk74Wz3JTHQ3u67HcleQ

**Pull request merged**